### ISSUE TITLE
Sampling T for Continuous Time

### ISSUE THREAD

**Issue by thorinf** [smsharma/consistency-models#1]
Hi,

I noticed in the continuous notebook that the `t_batch` is sampled uniformly, but that you were having issues with the training. I think one reason might be that the discretized time is not linearly spaced, it seems to be somewhat exponential. It might be useful to warp the continuous time to mimic this.

**Comment by smsharma** [comment]
Hi @thorinf, thanks for raising this. Can you expand a bit on what you mean by "the discretized time is not linearly spaced"? I was primarily going by Eqs. (30) and (33) of the [paper](https://arxiv.org/abs/2303.01469), where the expectation is taken over $t\sim\mathcal U[\epsilon, T]$, but I may well be missing something here. Indeed, as you say, the CT model at the moment does not seem to work.

**Comment by thorinf** [comment]
Sure, I've actually written a bit of code for it. So when you get the discretized karras boundaries the jumps aren't linear. These are the karras values for N=16:

`[2.0000e-03, 8.8270e-03, 3.0025e-02, 8.5087e-02, 2.1065e-01, 4.6998e-01,
        9.6542e-01, 1.8543e+00, 3.3686e+00, 5.8389e+00, 9.7232e+00, 1.5640e+01,
        2.4408e+01, 3.7092e+01, 5.5051e+01, 8.0000e+01]`

Whereas when sampling using continuous it is uniform sample, so the noise is generally going to be much larger in comparison.

Instead, it might be better to sample uniformly then warp them:

```python
def kerras_warp_scaling(u, sigma, eps, T):
    # u [0,1]
    return (eps ** (1 / sigma) + u * (T ** (1 / sigma) - eps ** (1 / sigma))) ** sigma
```
Discrete:
```python
rand_u_1 = torch.randint(N - 1, (img.shape[0],), device=img.device, requires_grad=False) / (N - 1)
rand_u_2 = torch.clamp(rand_u_1 + 1 / (N - 1), 0.0, 1.0)
```
Continuous:
```python
rand_u_1 = torch.rand((img.shape[0],), dtype=img.dtype, device=img.device, requires_grad=False)
rand_u_2 = torch.clamp(rand_u_1 + 1 / (N - 1), 0.0, 1.0)
```
Final Values for Discrete or Continuous:
```python
t_1 = kerras_warp_scaling(rand_u_1, 7.0, 0.002, 80.0)
t_2 = kerras_warp_scaling(rand_u_2, 7.0, 0.002, 80.0)
```
I've tried this in my code, and the model does seem to learn the basic shapes when I view images, however the loss is decreasing on epoch 0, then increases. It still produces progressively better images, but the loss is unstable. 

There are a couple of things that I could improve here:
- For the `rand_u_1` I could limit the value so it can't go right up to 1, instead `1 - 1 / (N - 1)` as max. But I don't feel that will make a huge amount of different.
- Instead of adding 1 for `rand_u_2` I could change it to a smaller value.

**Comment by thorinf** [comment]
Apologies for using Torch!

I am by no means saying this is the right way, but it does seem reasonable.

**Comment by thorinf** [comment]
I made an edit: i'd actually been using `randn` for `rand_u_1` instead of `rand`.  With the fix the model is stable in my implementation, no increasing losses - it looks like it will converge.
